ddtrace/internal/datadog/profiling/dd_wrapper/dd_wrapper.md

```diff
@@ -0,0 +1,74 @@
+# dd_wrapper
+
+This is experimental code that uses libdatadog for the collection and propagation of profiling data.  In theory this should just be a thin wrapper, but in practice some more state management.
+
+## Historical context
+
+Traditionally, dd-trace-py maintained three separate sampling mechanisms for the collection of profiling data.  These are called "collectors" in dd-trace-py parlance, but we'll refer to them as "samplers" in conformance with the broader litearture.
+
+### Stack sampler
+
+This is a simple stack sampler that collects stack traces at a dynamic interval.  Supposedly, it's a fixed-interval profiler, but the interval is adjusted dynamically depending on the observed overhead.  This sampler lives in a Python thread.  When it wakes up, it looks at the current threads/tasks and collects stack traces for each of them.  It also looks at the current exceptions and collects similar data for them.
+
+Although this sampler operates in its own thread, since its fundamental interaction requires taking the GIL, it never samples at a time when other threads are on-CPU (except if they're off-GIL in C code or in io or whatever, but that isn't an issue).
+
+### Lock sampler
+
+This patches some common lock implementations to perform sampling.  As such, the sampling occurs from the active application thread.
+
+### Memory sampler
+
+This is very similar to the lock profiler, except for allocations.
+
+
+## Current state
+
+For various reasons, we're validating profiling concepts where as much (or all!) of the work is done from native threads.  This is a bit of a departure from the traditional dd-trace-py model, but the technique is well-represented in the community.  This is comprised of two main parts right now
+
+* Moving the stack sampler to a native, off-GIL thread
+* Moving the collection apparatus to native code
+
+The problem here is that in Python it's fairly common to periodically call `fork()` as a worker lifecycle management technique.  When this happens, then in the child:
+
+* Execution resumes from the same point as the parent.  One way to think about this is, execution resumes from the thread that called fork
+* The heap is configured to be copied-on-write.
+
+This means any intermediate heap-stored state managed by a sibling thread (i.e., a thread which did not call `fork()`) may become corrupt in the child.  We need to protect or re-initialize such state when possible.
+
+## Data organization in this code
+
+Roughly, the data is organized in the following way
+
+### Samples
```

Comment:
It will emit two separate samples with identical stack traces.  Each individual string will get de-duplicated only one time (global storage), and each stack trace will get unwound separately.

There have been issues where this was a problem for some reason or another, but right now I think the overriding issue is a lost performance opportunity (e.g., only unwind once).